# Praktikum 4

In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf

Load data

In [2]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

In [5]:
print(dataset.head())

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

<p>Memuat dataset churn modeling dengan pandas, menampilkan 5 baris pertama dengan head. Variabel X diisi fitur dataset mulai dari kolom ketiga sampai satu kolom sebelum terakhir. </p>
<p>y berisi label target, nilai dari kolom terakhir dalam dataset</p>

In [4]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


Encoding data kategorikal

In [7]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])

In [8]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


Mengubah nilai kolom ketiga dengan labelEncoder.

Encoding Kolom "Geography" dengan One Hot Encoder

In [9]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))

In [10]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


Encoding menggunakan OneHotEncoder untuk encoding nama negara, kode ini menetapkan OneHotEncoder sebagai transformasi untuk kolom kedua, sedangkan kolom lainnya tetap tidak berubah berkat parameter remainder='passthrough'. Metode fit_transform kemudian diterapkan untuk mengubah data, menghasilkan array X yang telah dikonversi menjadi format numerik dengan representasi one-hot untuk kolom yang dikategorikan.

Split data

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

Scaling fitur

In [12]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Membuat model ANN

In [13]:
ann = tf.keras.models.Sequential()


Kode menginisialisasi model ANN menggunakan Keras, yang merupakan bagian dari tensofrlow. Objek ini bertipe sequential. Model ini memungkinkan penambahan lapisan-lapisan (layers) secara berurutan, di mana output dari satu lapisan menjadi input untuk lapisan berikutnya.

Mmebuat input layer dan hidden layer pertama

In [14]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))


<p>Menambahkan lapisan dense kedalam model ANN. Sebuah lapisan tersembunyi baru ditambahkan ke model dengan 6 unit neuron dan menggunakan fungsi aktivasi ReLU (Rectified Linear Unit).</p>



Mmebuat hidden layer kedua 

In [15]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))


Membuat output layer

In [16]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))


Fungsi aktivasi sigmoid menghasilkan nilai antara 0dan 1. Dalam konteks ini layer output akan memberi prediksi probabilitas untuk kelas target, dimana nilai mendekati 1 menunjukkan kemungkinan yang tinggi untuk salah satu kelas. sedangkan nilai mendekati 0 berkemungkinan rendah

## Training model

Compile model (menyatukan arsitektur ANN)

In [17]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])


konfiguras model ANN menggunakan metode compile. Paramter yang ditetapkan mencakup 
1. penggunaan optimizer Adam yang efsien dalam memperbarui bobot model selama proses training
2. loss binary_crossentropy yang membantu menghitung perbedaan antara prediksi model dan nilai target. Selain itu model siap dilatih menggunakan metode fit

Fitting model

In [18]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)


Epoch 1/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 780us/step - accuracy: 0.3922 - loss: 0.8877
Epoch 2/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 622us/step - accuracy: 0.7740 - loss: 0.6340
Epoch 3/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 618us/step - accuracy: 0.7898 - loss: 0.5876
Epoch 4/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 575us/step - accuracy: 0.8057 - loss: 0.5500
Epoch 5/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 632us/step - accuracy: 0.8091 - loss: 0.5255
Epoch 6/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 627us/step - accuracy: 0.8188 - loss: 0.4981
Epoch 7/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 547us/step - accuracy: 0.8217 - loss: 0.4787
Epoch 8/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 722us/step - accuracy: 0.8236 - loss: 0.4624
Epoch 9/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 706us/step - accuracy: 0.8373 - loss: 0.4410
Epoch 10/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 636us/step - accuracy: 0.8265 - loss: 0.4462
Epoch 11/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 565us/step - accuracy: 0.8294 - loss: 0.4328
Epoch 12/100
250/25

Menjalankan proses training untuk model ANN. Parameter batch_size=32 mengatur jumlah sampel yang akan diproses sebelum memperbarui bobot model, sementara epochs=100 menunjukkan bahwa model akan dilatih selama 100 iterasi penuh melalui dataset.

## Membuat prediksi

Diberikan informasi sebagai berikut,

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: $ 60000

Number of Products: 2

Does this customer have a credit card ? Yes

Is this customer an Active Member: Yes

Estimated Salary: $ 50000

Apakah customer tersebut perlu dipertahankan?

Memodelkan data baru dan buat prediksi

In [19]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 51ms/step
[[False]]


- Apakah hasilnya false?
  
<p>Hasil menunjukkan false berarti pelagnga ndengan karakteristik yang diberi tidak perlu dipertahankan. </p>
<p>Model ANN dilatih berdasarkan data historis, mungkin telah mengenali pola bahwa pelanggan dengan karakteristik serupa lebih cenderung untuk churn.</p>
<p>
Dengan threshold prediksi yang ditetapkan pada 0.5, nilai prediksi di bawah batas ini menunjukkan bahwa pelanggan tersebut diperkirakan tidak akan bertahan dengan perusahaan.
</p>

Prediksi dengan data testing

In [20]:

y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 583us/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


Model ANN digunakan untuk memprediksi potensi churn pelanggan pada data pengujian X_test. 
1. Prediksi dilakukan dengan memanggil metode predict() pada model ANN, yang menghasilkan nilai probabilitas untuk setiap sampel dalam X_test.
2.  Selanjutnya, nilai probabilitas tersebut diubah menjadi label biner dengan membandingkannya terhadap threshold 0.5; jika nilai probabilitas lebih besar dari 0.5, maka dianggap sebagai kelas 1 (pelanggan yang mungkin akan churn), dan jika tidak, dianggap sebagai kelas 0 (pelanggan yang tidak mungkin churn).
3. Kemudian, hasil prediksi (y_pred) dan label asli dari data pengujian (y_test) digabungkan dalam bentuk array dua kolom menggunakan np.concatenate(). Hasilnya adalah matriks yang menunjukkan label prediksi dari model bersamaan dengan label aktual, yang memungkinkan analisis lebih lanjut tentang akurasi dan performa model dalam mengklasifikasikan churn pelanggan.

Cek akurasi dan confusion matrix

In [21]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1505   90]
 [ 191  214]]


0.8595

Hasil evaluasi model ANN menunjukkan performa cukup baik.  di mana model berhasil mengidentifikasi 
- 214 pelanggan yang benar-benar churn (True Positives) 
-  1514 pelanggan yang tidak churn (True Negatives) 
-  terdapat 81 pelanggan yang salah diprediksi sebagai churn (False Positives) 
-   191 pelanggan yang sebenarnya churn tetapi tidak teridentifikasi oleh model (False Negatives). 
-   Akurasi keseluruhan model mencapai 86.4%,